internal/server/providers/oidc.go

```diff
@@ -0,0 +1,291 @@
+package providers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/go-playground/validator/v10"
+	"golang.org/x/oauth2"
+	"gorm.io/gorm"
+
+	"github.com/infrahq/infra/internal"
+	"github.com/infrahq/infra/internal/logging"
+	"github.com/infrahq/infra/internal/server/data"
+	"github.com/infrahq/infra/internal/server/models"
+	"github.com/infrahq/infra/uid"
+)
+
+const oidcProviderRequestTimeout = time.Second * 10
+
+// InfoClaims captures the claims fields from a user-info response that we care about
+type InfoClaims struct {
+	Email  string   `json:"email" validate:"required_without=Name"` // returned by default for Okta user info
+	Groups []string `json:"groups"`
+	Name   string   `json:"name" validate:"required_without=Email"` // returned by default for Azure user info
+}
+
+type OIDC interface {
```

Comment:
Not important for this PR, but we may want to rename this type since `oidc` is also a kind of provider

Reply:
Agreed, Im going to rename it `OIDCClient` or something similar. I'll cover that in my upcoming test coverage change.